signature-proof-spec/spec.md

```diff
@@ -0,0 +1,329 @@
+
+We will first start with a minimal Verifiable credential, then talk about how we protect the integrity of this document, while discussing this, we will also touch upon selective disclosure as a consequence of our implementation and its limitations. Lastly, we will extend our solution to allow a more efficient way of verifying credentials in a batch using merkle trees.
+
+We assume that the reader is familiar with basic cryptographic primitives and would not go into a deep discussion about them.
+
+First of all, lets start with a very minimal credential that adheres to the VC data model specification.
+
+```json
+{
+  "@context": [
+    "https://www.w3.org/2018/credentials/v1",
+    "https://www.w3.org/2018/credentials/examples/v1"
+  ],
+  "id": "http://example.edu/credentials/58473",
+  "type": [
+    "VerifiableCredential",
+    "AlumniCredential",
+    "OpenAttestationCredential"
+  ],
+  "issuer": "https://example.edu/issuers/14",
+  "issuanceDate": "2010-01-01T19:23:24Z",
+  "credentialSubject": {
+    "id": "did:example:ebfeb1f712ebc6f1c276e12ec21",
+    "alumniOf": "Example University"
+  },
+  "openAttestationMetadata": {
+    "template": {
+      "name": "EXAMPLE_RENDERER",
+      "type": "EMBEDDED_RENDERER",
+      "url": "https://renderer.openattestation.com/"
+    },
+    "proof": {
+      "type": "OpenAttestationProofMethod",
+      "method": "DOCUMENT_STORE",
+      "value": "0xED2E50434Ac3623bAD763a35213DAD79b43208E4"
+    },
+    "identityProof": { "identifier": "some.example", "type": "DNS-TXT" }
+  }
+}
+
+```
+
+The anatomy of the credentials is as above, but we will focus on the `credentialSubject` object as its values are the **set of claims** about the **subject**.
+
+For every property, nested or otherwise, we will first generate a secure random string (also known as a salt). This is to prevent a rainbow table attack [do i need to elaborate further?]. While traversing the key value pairs, we would also keep a reference to the path of the original structure of the credential subject.  After all the processing, we would flatten the values in an array of objects with the key value pairs of `value`, indicating the secure random string, and `path`, indicating the reference of the path of the original document.
+
+The code sample below shows the salted credential subject.
+
+```typescript
+const saltedCredentialSubject = [
+  {
+    value: "d775b5339c6b0adef3a2b8aaf33886b01759880ccad4e1f6cd4e44c18efadaca",
+    path: "credentialSubject.id",
+  },
+  {
+    value: "678465d280b40c984013af128913d30a4da311d5cfa75d4621efaa7921e1761a",
+    path: "credentialSubject.alumniOf",
+  },
+];
+```
+
+To further compact this output, we will first stringify the output from the salts, and then encode the stringified JSON in base64. This would be then inserted into the proof object as specified by the VC data model with the key of `salts`.
```

Comment:
As a note, We use base64 but we might consider other alternatives, as outlined in https://github.com/Open-Attestation/open-attestation/issues/141

Reply:
👌